File: 1to150/1_twoSum.java
class Solution{
    public int[] twoSum(int[] nums, int target){
        int[] ans = new int[2];
        if (nums == null || nums.length == 0){
            return ans;
        }
        //建立map 将所有int所对应的indx 存入hash中
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++){
            int thisNum = nums[i];
            map.put(thisNum,i);
        }
        
        for (int i = 0; i < nums.length; i++){
            int thisNum = nums[i];
            int remainder = target - thisNum;
            //如果该数的remainder存在于hash，表示我们找到了一pair，但还得确保这个数不会是自己，所以用indx来判断是不是
            //找到了自己，那么判断的依旧是 这个indx需要大于i
            if (map.containsKey(remainder) && map.get(remainder) > i){
                ans[0] = i;
                ans[1] = map.get(remainder);
            }
        }
        return ans;
    }
}

File: highFrequent/1_twoSum.py
'''
我们需要一个有记忆的机制，因为假设现在我们的数字是2， target是9， 那么我们要找的是7， 但现在我们不知道后面有没有7， 所以先把2存起来，如果遇到了7，则对于7来说它的remainder是2，2我们已经用记忆记住了，所以可以直接调取。

算法：
循环数组：
1. remain = target - now， 如果ramin出现在hash里调出这个remain的indx， return 结果
2. 直到循环结束return ans

'''
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        if (not nums):
            return []
        
        ans = []
        where = dict()        
        for i in range(len(nums)):
            now = nums[i]
            remain = target - now
            if remain in where:
                ans.append(where[remain])
                ans.append(i)
                return ans
            where[now] = i
        return ans
                
        
        